### DISCUSSION TITLE
Rancher Installation Error - cannot re-use a name that is still in use

### DISCUSSION THREAD

**Discussion by snarkipus** [mysticaltech/terraform-hcloud-kube-hetzner#401] [Q&A] [answered]
Fantastic tool - attempting to setup a dev server and I would like to install Rancher. I've set the following:
* control plane nodepool servertype: cpx21
*  initial_k3s_channel = "v1.24"
* enable_rancher = true
* rancher_hostname = "something.somewhere.io"
---
I'm not sure if I've misconfigured something, but the rancher install continuously fails:
```terminal
❯ kubectl get pods -n kube-system
NAME                                               READY   STATUS             RESTARTS        AGE
coredns-b96499967-llxs6                            1/1     Running            0               27m
hcloud-cloud-controller-manager-57d697cf98-jm7rw   1/1     Running            0               27m
hcloud-csi-controller-0                            5/5     Running            0               26m
hcloud-csi-node-95h4d                              3/3     Running            0               27m
hcloud-csi-node-zm5m7                              3/3     Running            0               27m
helm-install-cert-manager-mqsnr                    0/1     Completed          0               26m
helm-install-rancher-8f52q                         0/1     CrashLoopBackOff   9 (4m21s ago)   26m
kured-gmrml                                        1/1     Running            0               26m
kured-qjrk5                                        1/1     Running            0               26m
metrics-server-84f8d4c4fc-m4x5j                    1/1     Running            0               27m
traefik-864588c6c-fnbdp                            1/1     Running            0               26m
```
---
Pulling the logs from the failing install pod yields `Error: INSTALLATION FAILED: cannot re-use a name that is still in use`.
```terminal
❯ kubectl logs -n kube-system helm-install-rancher-8f52q | tail -20
Update Complete. ⎈Happy Helming!⎈
+ helm_update install --namespace cattle-system --repo https://releases.rancher.com/server-charts/stable
+ [[ helm_v3 == \h\e\l\m\_\v\3 ]]
++ ++ jq -r '"\(.[0].app_version),\(.[0].status)"'
helm_v3 ls --all -f '^rancher$' --namespace cattle-system --output json
++ tr '[:upper:]' '[:lower:]'
+ LINE=v2.6.9,uninstalling
+ IFS=,
+ read -r INSTALLED_VERSION STATUS _
+ VALUES=
+ for VALUES_FILE in /config/*.yaml
+ VALUES=' --values /config/values-01_HelmChart.yaml'
+ [[ install = \d\e\l\e\t\e ]]
+ [[ v2.6.9 =~ ^(|null)$ ]]
+ [[ uninstalling =~ ^(pending-install|pending-upgrade|pending-rollback)$ ]]
+ [[ uninstalling == \d\e\p\l\o\y\e\d ]]
+ [[ uninstalling =~ ^(deleted|failed|null|unknown)$ ]]
+ echo 'Installing helm_v3 chart'
+ helm_v3 install --namespace cattle-system --repo https://releases.rancher.com/server-charts/stable rancher rancher --values /config/values-01_HelmChart.yaml
Error: INSTALLATION FAILED: cannot re-use a name that is still in use
```
---
Looking at the overall deployment, I can only see the following in that namespace ... unfortunately it doesn't mean much to me as I'm not particularly k8s savvy.
```terminal
❯ kubectl get all --all-namespaces | grep -B 1 cattle-system
NAMESPACE        NAME                                                   READY   STATUS             RESTARTS      AGE
cattle-system    pod/rancher-post-delete-z54wb                          0/1     Completed          0             22m
--
NAMESPACE       NAME                                  COMPLETIONS   DURATION   AGE
cattle-system   job.batch/rancher-post-delete         1/1           49s        22m
```
Any help would be greatly appreciated. Thanks!

**Comment by mysticaltech** [comment, marked as answer]
@snarkipus First of all, please use this [kube.tf.example](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/multiple-fixes/kube.tf.example) (it will be released soon, but is compatible with the current release).

This problem showed up to me too in the past. Just make sure to use a proper real domain rancher hostname, not a dummy one, ideally you must own the domain, also point it to the lb.

Also, make sure to choose the rancher install channel of `latest`.

Hopefully, it works, but if not, please do include your kube.tf without the token and sensitive values (ideally with all comments meaning unused values removed).

**Reply by snarkipus** [reply]
@mysticaltech Thanks for the quick response!

> Just make sure to use a proper real domain rancher hostname, not a dummy one, ideally you must own the domain, also point it to the lb.

I just want to clarify, I should set:
`rancher_hostname = "newranchersubdomain.myrealdomain.com"`

Now I'm getting the following, which I assume is because it's trying to resolve the IP from cloudflare tied to "myrealdomain.com".
```
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization: Still creating... [40s elapsed]
...
╷
│ Error: remote-exec provisioner error
│
│   with module.kube-hetzner.null_resource.kustomization,
│   on .terraform/modules/kube-hetzner/init.tf line 231, in resource "null_resource" "kustomization":
│  231:   provisioner "remote-exec" {
│
│ error executing "/tmp/terraform_30196683.sh": Process exited with status 124
╵
```
---
Is there a DNS record/redirect I need to add? Feels like a chicken-and-egg situation since I need to redirect the subdomain to the load balancer (via an `A` record I assume) which requires the IP ... I feel like I'm missing a step here.

### Config File:
[kube.tf..txt](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/files/9944168/kube.tf.txt)

**Reply by mysticaltech** [reply]
@snarkipus Easy fix, cpx11 is too little (memory) for rancher (as explained in the docs), just change it to cx21 and it should work!

**Reply by snarkipus** [reply]
@mysticaltech Unfortunately, changing the `agent_nodepools` to `cpx21` didn't have any effect and the `control_plane_nodepools` was already `cpx21`. 
>IMPORTANT: Rancher's install is quite memory intensive, you will require at least 4GB if RAM, meaning cx21 server type (for your control plane).

Not exactly sure what acts as the DHCP server for this ... I assume Hetzner assigns the IP. Obviously the FQDN (i.e. newranchersubdomain.myrealdomain.com)  for the load balancer interface won't resolve until I update the DNS records though. I'll keep researching, maybe it's something silly.

Looks related to this issue #397 - Maybe #404 will resolve it.

**Reply by mysticaltech** [reply]
Ok, have decided to give it a shot thanks to your kube.tf, and am seeing the same. Currently investigating.

**Reply by mysticaltech** [reply]
@snarkipus It was the taint, just remove it and it then works. 

This one

      taints = [
        "server-usage=storage:NoSchedule"
      ],

replace it with

      taints = [],

**Reply by mysticaltech** [reply]
But before, destroy, `terraform upgrade -init`, and apply again with control plane node of at least cx21!

_Just FYI, about the upcoming PR #404, it won't affect your use case, so you are all good. But anyways, it will probably be merged very soon (like in the next 24h)._

**Reply by snarkipus** [reply]
Thanks @mysticaltech ! That sorted it ... now I just need to figure out how to get to the Rancher desktop!

Thank you so much for your help!

For anyone else that stumbles across this:
* ` kubectl get pods -A -o wide` to get the rancher pod install name
* `kubectl logs -n kube-system helm-install-rancher-nhkxd` to pull the logs from the successfully completed install

I had to do the next step in 2 parts: 
1) `kubectl get secret --namespace cattle-system bootstrap-secret -o go-template='{{.data.bootstrapPassword|base64decode}}'` to get the bootstrap password
2) enter `https://yoursubdomain.somwhere.com/dashboard/?setup=bootstrappasswordfrompreviousstep` to get to the dashboard

**Reply by mysticaltech** [reply]
If your DNS is on Cloudflare, just activate orange proxy button in your zone editor for that A record, that you point to the LB IP (you can see it with `kubectl get svc -A`). 

And go to SSL, choose "strict", to allow the custom rancher certificate on origin! And then just visit the link on https://....

Enjoy! :)

**Reply by snarkipus** [reply]
Hmmm ... interesting, proxying through Cloudflare and setting `strict` gets me an invalid SSL certificate. I'll keep poking at it.

Update: @mysticaltech I ended up just settling on `Full` since (as far as I can tell) using the Rancher generated CA would require using 'Advanced Certificates' from Cloudflare ($10/mon). I'm not entirely sure if that's expected/known. I suppose the other option would be to use Cloudflare's certs, but this should work for a non-production environment (I'm guessing).

**Reply by mysticaltech** [reply]
My bad, it's the "full" that probably allows the custom certificate on origin. Apologies!
